compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp

```diff
@@ -335,6 +335,164 @@ struct AllReduceOpConversion : public OpConversionPattern<mhlo::AllReduceOp> {
+
+  // Reshape
+  const int64_t rank = inputShape.size();
+  llvm::SmallVector<int64_t> newShape;
+  for (int64_t i = 0; i < rank; ++i) {
+    if (i != splitDim) {
+      newShape.push_back(inputShape[i]);
+      continue;
+    }
+    newShape.push_back(splitCount);
+    newShape.push_back(inputShape[i] / splitCount);
+  }
+  Value result =
+      rewriter
+          .create<mhlo::ReshapeOp>(
+              loc, RankedTensorType::get(newShape, inputType.getElementType()),
+              input)
+          .getResult();
+
+  // Transpose
+  SmallVector<int64_t> permutation;
+  permutation.reserve(rank + 1);
+  for (int64_t i = 0; i < rank; ++i) {
+    int64_t dimAfterReshape = i >= splitDim ? i + 1 : i;
+    if (i == concatDim) {
+      permutation.push_back(splitDim);
+    }
+    permutation.push_back(dimAfterReshape);
+  }
+  result = emitTranspose(rewriter, loc, result, permutation);
+
+  // Reshape
+  llvm::SmallVector<int64_t> finalShape(inputShape);
+  finalShape[concatDim] *= splitCount;
+  finalShape[splitDim] /= splitCount;
+  return rewriter
+      .create<mhlo::ReshapeOp>(
+          loc, RankedTensorType::get(finalShape, inputType.getElementType()),
+          result)
+      .getResult();
+}
+
+struct AllToAllOpConversion : public OpConversionPattern<mhlo::AllToAllOp> {
+  using OpConversionPattern<mhlo::AllToAllOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::AllToAllOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+
+    // Create a channel.
+    auto channel = rewriter.create<IREE::Flow::ChannelDefaultOp>(
+        loc, /*group=*/StringAttr{});
+
+    // Get the collective element type attribute.
+    auto resultType = op.getResult(0).getType().cast<RankedTensorType>();
+    IREE::Flow::CollectiveElementTypeAttr elementTypeAttr =
+        getCollectiveElementTypeAttr(op.getContext(), resultType);
+    if (!elementTypeAttr) {
+      return rewriter.notifyMatchFailure(
+          op, "unsupported element type for collective op");
+    }
+    if (op.getNumOperands() != 1) {
+      return rewriter.notifyMatchFailure(op,
+                                         "tuple all-to-all is not supported");
+    }
+    if (!op.getSplitDimension() || !op.getConcatDimension() ||
+        !op.getSplitCount()) {
+      return rewriter.notifyMatchFailure(
+          op,
+          "split_dimension, concat_dimension, and split_count must be present "
+          "for array all-to-all");
+    }
+
+    uint64_t splitDim = *op.getSplitDimension();
+    uint64_t concatDim = *op.getConcatDimension();
+    uint64_t splitCount = *op.getSplitCount();
+    Value allToAllInput = op.getOperand().front();
+    ArrayRef<int64_t> inputShape =
+        allToAllInput.getType().cast<RankedTensorType>().getShape();
+    SmallVector<int64_t> allToAllResultShape(resultType.getShape());
+
+    const bool requiresRearrange = concatDim != splitDim;
```

Comment:
IIUC, we need pre-transpose and post-transpose when `split_dim != 0`.
Also, we don't need a rearrangement before all_to_all and only need it when `concat_dim != split_dim`. Here is the logic.
1. transpose input if `split_dim != 0`.
2. do all_to_all
3. transpose output if `split_dim != 0`.
4. rearrange if `concat_dim != split_dim`.

Please verify this is right. If this is right, it would make the code simpler.

Also, reuse the transpose logic I made for all_gather. We can further simplify the logic by putting the permutation vector creation inside the function. In our case, we always transpose with two dimensions, so `emitTranspose(input, srcDim, dstDim)` might be the simplest.

`splitAndConcat` may be clearer than `rearrange`.

Reply:
Thanks for the review @okkwon! 

Let me try to simplify this logic and reuse your transpose code as well.

Reply:
@okkwon Just pushed the changes. Please let me know what you think. Thanks!

Reply:
Looks much better and easy to understand. Thanks!